### DISCUSSION TITLE
AVR DD vs TINY 2

### DISCUSSION THREAD

**Discussion by trimarco232** [SpenceKonde/DxCore#293] [Show and tell]
..........................................................AVR16DD14.........ATTINY1624
General Purpose I/O ........................9(10,11) ....... 11(12)

**Comment by mechatroniks-git** [comment]
Nice comparison. Can you add code formatting so everything lines up?

**Comment by trimarco232** [comment]
Hi,
no, it doesn't work, don't now why, maybe my browser

**Comment by fsievers** [comment]
I just transferred it into a table.

Spence: I edited the table to clear up a few things and add a bit more information where I thought key things were omitted. 
Basically, the Dx oscillator tuning is not useful to get far from the center frequency. The tinyAVR oscillator has a mindblowing compliance. 20MHz base can be tuned from  just under 14 to over 35 at which point its' not at 127 yet, that's just where the chip stopped executing instructions correctly ;-)

| | AVR16DD14 | ATTINY1624 |
|---:|-------------------|---|
Pins | 14 | 14
Flash Memory | 16KB | 16KB
SRAM | 2KB | 2KB
EEPROM | **256B** | 128B
Maximum flash announced | **64KB** | 32KB
SRAM of largest-flash version | **8k** | 3k
Max. *rated( frequency (MHz) | **24** | 20
*Internal oscillator has meaningful turning | No | **Yes**
*Internal oscillator max speed | **32 Mhz w/out tuning, just set bit field to 0x0B** | 32 MHZ tuned aggressively*
External HF crystal | **Yes** | No
External HF Xtal overclock | likely 40 MHz | n/a 
External clock overclock    | likely 48 MHz | maybe 36? 
16-bit Timer/Counter type A (TCA) | 1 | 1
16-bit Timer/Counter type B (TCB) | 2 | 2
12-bit Timer/Counter type D (TCD) | **1** | no
Real-Time Counter (RTC) | 1 | 1
USART | 2 | 2
SPI | 1 | 1
TWI / I2C (dual mode) | **1(1)** | 1
PORTMUX optiions for serial interfaces | **extensive** | limited
PORTMUX options for TCA0 PWM | per port `***` | **per pin** 
12-bit differential ADC (channels) | kinda-sorta 1 (7) | **1(9)** `*`
10-bit DAC (outputs) | **1(1)** | 0
Analog Comparator (AC) | 1 | 1
Zero-Cross Detector (ZCD) | **1** |0
Configurable Custom Logic (CCL LUT) | 4 | **4**
Event System (EVSYS) channels | 6 | 6
Multi-Voltage I/O (MVIO) | **YES** | NO
General Purpose I/O | 7-11 `**` | **11(12))**
1—24 | €0.90 | **€0.83**
25—99 | €0.82 | **€0.77**

* Notes from Spence

`*` Calling what the DD has a differential ADC is a real stretch. Both voltages must be within ground and the reference. Its as if you just had 2 ADCs and kicked them off at the same time and subtracted. On 2-series and upcoming EA series, it just needs to be between gnd - a fraction of a volt and Vcc + a fraction of a volt, so you can do high side current sensing and cool stuff like that. 
The 2-series also has a programmable gain amplifier, the DD does not have anything like that, performs conversions faster, and can automatically accumulate 1024 measurements, which you can then decimate to measure with 17 bits of accuracy. Now how many of those bits are noise vs signal, well, that depends on how stable the reference and power supply decoupling in a major way. The PGA can also amplify single ended measurements!

`**`A DD14 can have at most 10 I/O pins and 1 input only pin, because it can't change the fact that it's got an extra power pin. 
 The input pin can be gotten without penalty just by sacrificing reset. the 3 remaining GPIOs are the two crystal ones, which you may well want or need for a crystal - it's 9 without the crystal. And finally if you have an HV programmer, you can configure UPDI to act as GPIO, and reset and input, and you now have 11 opins and have lost most of the features you probably wanted the chip for. 

Note on LUTs: the 2-series lacks rhyme or reason to placement of CCL pins (it's classic Atmel dartboard style). On the other hand, the DD-series orderly approach means that low pincount devices just don't have hardly any LUT input pins available. Not that the lut pin inputs arewhat anyone cares about - the CCL is about the INTERNAL inputs, which allow you do to mighty and terrible things; IMO the combinantion of EVSYS and CCL is the most powerful and flexible peripheral on the modern AVRs. 

`***` Per-port mapping of TCA0 is much better for with 32+pin parts, but inferior with smaller pincounts. 
The tinyAVR 2-series has 2 USARTs, each of which can use 1 pin set. but the default USART1 pinset and alternate USART1 pinset are shared, two SPI pinsets (both of which share pins with the pins that can be used with USART1. The 2-series also doesn't permit remapping the SCL and SDA pins of the TWI interface (though TWI, even without dual mode has been able to act as both master and slave for a while now on the same pins (multimaster bus) - though the dual mode is usually greatly preferable), and those pins are the same as USART0's XDIR and XCK pins. The pin mapping options for serial interfaces available on the DD's are really some next level shit as the kids say now. 

make sure you read the portmux section, They clearly added them for lowpincount versions, but high pincount parts get them too.... and that may be the standard moving forward!

**Reply by SpenceKonde** [reply]
I added the DD and future EA-series parts to the big comparison table in megaTinyCore readme.

**Reply by ObviousInRetrospect** [reply]
I think the General Purpose I/O line has a bug - not sure how you get 13 on a 14-pin part. The data sheet just says 12.

**Comment by trimarco232** [comment]
right, sorry
un addition, how one can use the special pins (RESET, UPDI) as GPIOs isn't that obvious

**Comment by SpenceKonde** [comment]
Will correct. Should be 11 at the most, one of which is input only and
requires disabling reset, and another of which requires disabling updi
(which there has normal drive strength). He pulse on dd also needs to be
delivered to reset.

Something about the hb tolerance ce make output drivers terrible

I'm nit immediately seeing whet page this is on, can you point me in the
general direction?   Thanks

Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by ObviousInRetrospect** [reply]
actually page 2 of https://ww1.microchip.com/downloads/aemDocuments/documents/MCU08/ProductDocuments/DataSheets/ATtiny3224-3226-3227-Data-Sheet-DS40002345B.pdf says 12.

page 48 says the GPIO output driver is disabled for 768 OSC32K cycles after system reset.

So the

**Reply by SpenceKonde** [reply]
Yeah, that is correct on tinyAVRs only, this is because there is no separate pin to put the HV pulse to override the fuse settings and turn the pin back into UPDI. Hence, when that pin is used as GPIO, a "PCHV" cycle is required to to start programming (power-cycle, then hv pulse, then program normally). It's much easier on the DD's , where there's a separate reset pin, and it can get the HV pulse. That's something you can do with a pair of fets. But because the UPDI pin is connected to your programmer, you need a SPDT electrically controlled switch  with guaranteed BBM to do a PCHV programming cycle for a tiny, to avoid applying 12v to the pins of your programmer, and destroying the chip in short order.  I'm working on such a programmeer (to be called HyperUPDI due to the large speed advantage it will also offer, particularly for tinyAVR, by using a more efficient protocol between the computer and the programmer than updi, with fewer USB latency delays), which will have UPDI, HV UPDI, selectable 5V0, 3V3, or V<sub>target</sub> programming voltage. In all except the last case, where the target has it's own power during programming and the Vcc line is only connected to the VDDIO2 line of the on-board microcontroller - and when not programming will act like a normal serial adapter). Naturally this will cost considerably more than throwing a diode and/or resistor onto a 99 cent serial adapter :-P  (that programmer will have a followon that will also support standalone programming, storing flash images and fuses in external SPI EEPROM or flash memory

**Comment by SpenceKonde** [comment]
It's. Fuse setting like usual (its exposed by the core wirh warning if the
option will require hv to do firther programming . If you set reset as inpu
t, you just lose the reset function. Updi still works. If you set updi as
gpio, you can t do any updi programming without an Hv programmer that
understands that dd series and later gets they hv pulse on reset, not updi.
I expect to have one for sale late summer

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Comment by ObviousInRetrospect** [comment]
Noticed what might be a significant difference for some applications that might be worth putting in this table:

The voltage references in the DD are all a horrendous +/- 4% [1]  instead of the +/-2% [2] in the attiny3224 for the 1v024 and 3% for the others. I went back and checked the  attiny1614 which also specifies 2% for 1v024. When I went to check the other DX parts I couldn't find any spec for the vref in the da/db (?!?!). I'd be kind of surprised if these were also 4%, does anyone have a number for the DA/DB? Edit: it isn't in the TOC, but DS40002300A-page 611 does in fact say that all the v-refs on the DB are "typical +/-4" with no min or max specified. Yuck. Looks like I may start populating an external vref or using vcc more often. I usually target tighter vcc regulation than +/-4.

[1]https://ww1.microchip.com/downloads/aemDocuments/documents/MCU08/ProductDocuments/DataSheets/AVR32-16DD20-14-Prel-DataSheet-DS40002413.pdf page 575
[2]https://ww1.microchip.com/downloads/en/DeviceDoc/ATtiny3224-3226-3227-Data-Sheet-DS40002345A.pdf page 489

**Reply by SpenceKonde** [reply]
The ADC on the DD was IMHO a stopgap measure because they didn't have the new ADC ready and they needed to ship a differential ADC on modern parts. As they were preparing the 1-series parts for 16 and 32k release, they realized that the differential ADC for the Dx-series (what the 2-series and EA got) was going to be way later than they wanted. So what do they do? They don't have much time, but they do have time to put a second copy of the ADC onto the larger 1-series parts, thus giving customers a way to simulate differential ADC only without the main advantages that a proper differential ADC has, and also gave them a testbed to see if that worked well enough that they could ship something derived from ti on the Dx-series. So you'd set one ADC to read the positive pin, the other to read the negative one, then link them both to the same generator channel and kick them off simultaneously and subtract the results to get a differential-like ADC measurement. (In that way they would have a testbed for a similar system - which they did a lot on the 1-series. First appearance of the TCD, the triple comparator contraption, not to mention the half-baked event system that the 1-series got. They definitely realized that they'd fucked up that pretty bad (come on, sync and async channels? with completely different organization of the generators? And the position of some peripherals within an async channel depending on which channel?!) - but had the current version working in time for the megaAVR 0-series; it's getting changed again for EA-series, to make all the channels interchangeable (thank god) - each port, and the RTC, will be given 2 channels to the event system and a single 8-bit register split into 2 4--bit bitfields for the two generators. Makes things simpler in some ways, but more complex in others. ). Considering the limitations of differential ADC on the Dx-series, I think they just do the same thing just more tightly integrated, without any of the analog wizardry they do on the 2-series that lets them use a 1v reference to see how far apart 2 voltages up at 5.0 and 4.9xx volts  are with 16x amplification.

**Comment by SpenceKonde** [comment]
Re reference accuracy - it's worth testing how accurate they are under realistic conditions, because the figures they give are often over the whole temperature range, whereas we often know that our devices will be operated in a relative ly narrow temperature band

**Comment by ObviousInRetrospect** [comment]
Yeah the ADC not being ready is one thing. The references being massively less accurate than dividing down VCC off the cheap LDOs I use that I select for low quiescent current and not accuracy is still unacceptable and seems unrelated.

I was pleasantly surprised to get a response from Microchip on my support case on the DB datasheet (they have been pretty good with documentation cases pointing out obvious documentation errors and asking for clarifications). They are tightening up the spec somewhat (in the direction of an O(10%) better being the min/max instead of the typical) but it is still disappointing for what seems like their analog flagship (built in op-amps etc). You'll note the tiny-2 has somewhat more restricted temperature ranges for the 2% spec. I'd be happy(er) with something like 1% typ at 25c and +/-2 max from 0c-40c. But really for a 12-bit ADC I want 0.5% typical, 1% min/max over temperatures humans are comfortable in for at least one reference. Can usually divide down the input to use the good reference. Next rev of my dev board for this chip is getting a footprint for a non-terrible external vref.

Also, I think the new docs say don't use Fclkadc >500khz with the 1.024v reference on the DB but it still wasn't clear to me whether that applies to an external reference or the internal references too. 

I didn't ask explicitly but I kind of expect all of the above will apply to the DD/DA as well.
